[feat] Add sparse index support to knowhere

zhengbuqian opened the pull request:
/kind feature
#193 

This PR adds the basic framework of supporting sparse index in knowhere, with 2 index types: inverted index and wand. 

In this PR only Search is supported. 

Support for growing segment is sub-optimal, we naively double the underlying data structure each time we used up the current space.

Added basic unit tests:
* Search without filtering, with different dataset sizes/recall/sparisity, etc.
* Search with filtering, with different filter rate.

alexanderguzhva commented on include/knowhere/utils.h:
```diff
@@ -92,4 +94,155 @@ readBinaryPOD(R& in, T& podRef) {
+    }
+    table_t
+    pop() {
+        std::pop_heap(pool_.begin(), pool_.begin() + size_--);
```
Please
```
std::pop_heap(pool_.begin(), pool_.begin() + size_);
size_ -= 1;
```

alexanderguzhva commented on src/common/comp/brute_force.cc:
```diff
@@ -343,4 +343,96 @@ BruteForce::RangeSearch(const DataSetPtr base_dataset, const DataSetPtr query_da
+                        dist += it->second * data[k];
+                    }
+                }
+                if (dist > 0) {
```
the code does not distinguish two very different situations:
1) `dist=0` because no item was found in `query`
2) `dist=0` because the match is exact

alexanderguzhva commented on src/index/sparse/sparse_inverted_index.h:
```diff
@@ -0,0 +1,449 @@
+        }
+        std::sort(q_vec.begin(), q_vec.end(),
+                  [](const auto& lhs, const auto& rhs) { return std::abs(lhs.second) > std::abs(rhs.second); });
+        while (q_vec.size() && q_vec[0].second * drop_ratio_search > q_vec.back().second) {
```
`while (q_vec.size() && ...)` -> `while (!q_vec.empty() && ...)`

alexanderguzhva commented on src/index/sparse/sparse_inverted_index.h:
```diff
@@ -0,0 +1,449 @@
+
+ private:
+    [[nodiscard]] float
+    dot_product(std::unordered_map<IndicesT, T> q_map, table_t u) const {
```
`dot_product(const std::unordered_map<IndicesT, T>& q_map, table_t u) const {`

alexanderguzhva commented on src/index/sparse/sparse_inverted_index.h:
```diff
@@ -0,0 +1,449 @@
+    }
+
+    void
+    refine_and_collect(std::unordered_map<IndicesT, T>& q_map, MinMaxHeap<T> inaccurate, size_t k, float* distances,
```
`const std::unordered_map`

alexanderguzhva commented on src/index/sparse/sparse_inverted_index.h:
```diff
@@ -0,0 +1,449 @@
+        if (refine_factor == 1) {
+            collect_result(heap, distances, labels);
+        } else {
+            std::unordered_map<IndicesT, T> q_map;
```
the size of the `q_map` is known beforehand (it is `len`). So, please construct `q_map` with the provided number of buckets (I'd try `4 * len` as a start), so 
```
std::unordered_map<IndicesT, T> q_map(4 * len);
```
The number of buckets impacts the performance of the following `dot_product()`

alexanderguzhva replied:
Or I can tweak the performance as soon as this commit lands, there are many things to play with there :)

alexanderguzhva commented on src/index/sparse/sparse_inverted_index.h:
```diff
@@ -0,0 +1,449 @@
+
+    template <typename HeapType>
+    void
+    collect_result(HeapType heap, float* distances, label_t* labels) const {
```
`HeapType& heap`

alexanderguzhva commented on src/index/sparse/sparse_inverted_index.h:
```diff
@@ -0,0 +1,449 @@
+            }
+        }
+        Cursor(const Cursor& rhs) = delete;
+        Cursor(Cursor&& rhs) noexcept {
```
Won't `Cursor(Cursor&& rhs) noexcept = default;` work?
Same question for `Cursor& operator=(Cursor&& rhs) noexcept`

alexanderguzhva commented on src/index/sparse/sparse_inverted_index.h:
```diff
@@ -0,0 +1,449 @@
+        size_t num_vec_ = 0;
+        float max_score_ = 0.0f;
+        float q_value_ = 0.0f;
+        const BitsetView* bitset_;
```
BitsetView is a lightweight structure (a wrapper around the pointer, basically) that should be passed by reference

alexanderguzhva commented:
the functionality seems to be fine, but what about the unit tests running time, is it expected?

hhy3 commented on src/index/sparse/sparse_inverted_index.h:
```diff
@@ -0,0 +1,430 @@
+
+ private:
+    [[nodiscard]] float
+    dot_product(const std::unordered_map<IndicesT, T>& q_map, table_t u) const {
```
std::unordered_map seems with a bad performance as I tested. It's better to use two arrays `indices` and `data`

zhengbuqian commented:
> the functionality seems to be fine, but what about the unit tests running time, is it expected?

I added some large unit tsets so ut took quite long, I have removed them. Will revert the change in ut yaml before merge.

zhengbuqian commented:
/hold

zhengbuqian commented:
/unhold

zhengbuqian commented:
/run-e2e

zhengbuqian commented:
run-e2e

alexanderguzhva commented:
/lgtm overall

samhuang-z commented:
/run-e2e

samhuang-z commented:
/run-e2e

zhengbuqian commented:
/hold

zhengbuqian commented:
Use https://github.com/zilliztech/knowhere/pull/341 instead
